### PR TITLE
Add Enterprise SONiC "Vlan Mapping" Resource Module

### PR DESCRIPTION
**Pull request by Cypher-Miller**
**SUMMARY**

- Add Enterprise SONiC 'vlan mapping' resource module:
  -  Provide data definitions in 'argspec'.
  - Implement 'facts' to gather facts for the module.
  - Implement 'config' to apply configurations for merged, deleted, overridden, and replaced states.
- Define regression test cases to verify correct functionality of vlan mapping resource module.

**ISSUE TYPE**
New Module Pull Request

**COMPONENT NAME**
sonic_vlan_mapping

**ADDITIONAL INFORMATION**

**Test Results**

- Regression test HTML report:
 [vlan_mapping_regression.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/11080566/vlan_mapping_regression.pdf)

- Regression test detailed log:
[regress_output.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/11056338/regress_output.txt)

- Show running config and respective facts gathered:
[vlan_mapping_show_running_config.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/11080998/vlan_mapping_show_running_config.txt)

### CODE REVIEW

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+        :rtype: A list
+        :returns: the commands necessary to migrate the current configuration
+                  to the desired configuration
+        """
+        requests = []
+        commands = []
+        commands_del = []
+
+        if diff:
+            commands_del = self.get_replaced_delete_list(want, have)
```
Comment (mingjunzhang2019) [inline plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:151]: When a mapping M is in both want and have, the attributes in want.M have same values as those attribues in have.M, but want.M has less attributes than have.M, so diff=[]. The result should be that have.M is replaced by want.M? Please verify.

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+        if commands_del:
+            requests_del = self.get_delete_vlan_mapping_requests(commands_del, have, is_delete_all=True)
+            if requests_del:
+                requests.extend(requests_del)
```
Comment (mingjunzhang2019) [inline plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:155]: "command_del" should be "update_states" and "extend" into "commands".

Comment (Cypher-Miller) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:155]: Are you saying I should add lines similar to this but for the delete commands? 
commands_over = update_states(commands_over, "overridden")
commands.extend(commands_over)

File: plugins/module_utils/network/sonic/facts/vlan_mapping/vlan_mapping.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
+#
+# -*- coding: utf-8 -*-
+# Copyright 2022 Dell Inc. or its subsidiaries. All Rights Reserved
```
Comment (stalabi1) [inline plugins/module_utils/network/sonic/facts/vlan_mapping/vlan_mapping.py:3]:
```suggestion
# Copyright 2023 Dell Inc. or its subsidiaries. All Rights Reserved
```

File: plugins/modules/sonic_vlan_mapping.py [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+#interface Ethernet16
+# mtu 9100
+# speed 400000
+# fec RS
+# unreliable-los auto
+# shutdown
+# switchport vlan-mapping 400-402,404,406,408,410,412,420,422,430-431 dot1q-tunnel 2436
+#!
+
+  - name: Overwrite the vlan mapping configurations
```
Comment (stalabi1) [inline plugins/modules/sonic_vlan_mapping.py:448]:
```suggestion
  - name: Override the vlan mapping configurations
```

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py (hunk 2 above)
Comment (mingjunzhang2019) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:155]: Something like this: commands.extend(update_states(command_del, "deleted"))

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py (hunk 1 above)
Comment (Cypher-Miller) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:151]: After doing some testing I see what you mean the "if diff:" is causing the commands_del to not run when it should so I have removed it and checking the changes now.

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py (hunk 2 above)
Comment (Cypher-Miller) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py:157]: Pushed the changes

File: plugins/module_utils/network/sonic/facts/vlan_mapping/vlan_mapping.py (hunk 3 above)
Comment (Cypher-Miller) [reply plugins/module_utils/network/sonic/facts/vlan_mapping/vlan_mapping.py]: Pushed the changes

File: plugins/modules/sonic_vlan_mapping.py (hunk 4 above)
Comment (Cypher-Miller) [reply plugins/modules/sonic_vlan_mapping.py]: Pushed the changes

File: plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py (hunk 1 above)
Comment (Cypher-Miller) [reply plugins/module_utils/network/sonic/config/vlan_mapping/vlan_mapping.py]: I have pushed the fix for this. I had to make some more changes for errors that occurred after removing the diff check.

**Review by mingjunzhang2019 — APPROVED** [review]

**Review by stalabi1 — APPROVED** [review]

**Review by kerry-meyer — APPROVED** [review]

**Pull request merged**